improved the checks if a planned transfer is possible or not

RndName opened the pull request:
fix for #1454 

- instead of only checking if the transfer destination was captured it now checks if there is a valid route between origin and destination. This also ensures that there will be a check if the current position or next_stop was captured and therefore the transfer should be disbanded.
- disband uncompletable transfer before planning or performing (also when user cheated a base capture)

Unfortunately the function to find an escape route is not implemented yet. But having this function some time implemented will make the possible check way better. #1457 will be there for it.

This PR can of course be cherry-picked to V4 without any problems

RndName committed:
improved the validation for planned transfers

- instead of only checking if the transfer destination was captured it now checks if there is a valid route between origin and destination. This also ensures that there will be a check if the current position or next_stop was captured and therefore the transfer should be disbanded.
- disband uncompletable transfer before planning or performing (also when user cheated a base capture)

DanAlbert commented on game/transfers.py:
```diff
@@ -166,9 +166,9 @@ def find_escape_route(self) -> Optional[ControlPoint]:
-    def proceed(self) -> None:
-        if not self.destination.is_friendly(self.player):
-            logging.info(f"Transfer destination {self.destination} was captured.")
+    def possible(self) -> bool:
```
The name of this function makes it sound like it shouldn't have side effects. Alternate refactor:

```
def should_disband(self) -> bool:
    ...

def disband(self) -> bool:
    ...
```

The former ends up pretty simple, and the latter de-dups a lot of what's here, while still decluttering `proceed`, I think?

RndName replied:
agree! i added both as i removed some duplication in the logging as well

DanAlbert commented on game/transfers.py:
```diff
@@ -166,9 +166,9 @@ def find_escape_route(self) -> Optional[ControlPoint]:
-        if not self.destination.is_friendly(self.player):
-            logging.info(f"Transfer destination {self.destination} was captured.")
+    def possible(self) -> bool:
+        if self.transport and not self.next_stop.is_friendly(self.player):
```
Disbanding should happen even if `self.transport is None`, otherwise airlifts that were waiting for aircraft will never be pruned.

RndName replied:
thanks i was already wondering why transport could be none! i added the check for transport none and destination is not friendly as an airlift only has a destination

DanAlbert replied:
I think `next_stop` being `None` shouldn't cause the _transfer_ to cancel, since there might be an alternative route. With this change the whole transfer will cancel if the _best_ route is cutoff.

This is aiming to fix the case where the `next_stop` was captured, but the destination was still available, right? I think for that case `process` should check that the next stop is invalid, make no progress that turn, and allow the replan. Probably will need to add a check in `PendingTransfers.arrange_transport` to cancel the transfer when there is _no_ path between the current location and the destination, but that should be rare.

RndName replied:
@DanAlbert  i changed the logic behind the checks a bit to first check if the transit network has a valid link between `position` and `destination` once the `next_stop` gets captured. 
The current logic of the game always calculates the optimum route during `arrange_transport` and creates a transport for the route `transfer.position` to `transfer.next_stop` or `transfer.destination` depending on the possible `link_type`. 
So i only check for `next_stop` in the validation and set `transport` to `None` whenever the `next_stop` gets captured and a link in the transit network still exists. As the `arrange_transport` will be triggered afterwards it will again calculate an optimum route and will setup a new transport. In my testing it mostly just setups an airlift instead because no road network is available anymore. But i think its really rare that there will be an alternative road network as this needs a really complicated conflict theater with lots of CPs. But in total the logic of the code can handle it :)

I think i covered any possibility with this. May i have overseen sth?

DanAlbert replied:
> In my testing it mostly just setups an airlift instead because no road network is available anymore.

Yeah, that should be typical for most campaigns.

> But i think its really rare that there will be an alternative road network as this needs a really complicated conflict theater with lots of CPs.

These complicated cases were a lot more common before. Those scenarios never really worked very well so when all the old campaigns were tossed out with 3.0 none of the new ones recreated those complicated networks.

DanAlbert commented on game/transfers.py:
```diff
@@ -166,21 +166,49 @@ def find_escape_route(self) -> Optional[ControlPoint]:
+            )
+            self.kill_all()
+
+    def should_disband(self, network: TransitNetwork) -> bool:
```
This has the same naming confusion I mentioned before. This function shouldn't have any side effects. The _caller_ should be the one calling `disband`, not this function.

In other words, `validate_transfers` should be:

```
def disband_uncompletable_transfers(self) -> None:
    remaining = []
    for transfer in self.pending_transfers:
        if transfer.should_disband(self.network_for(transfer.position)):
            transfer.disband()
        else:
            remaining.append(transfer)
    self.pending_transfers = remaining
```

It makes `validate_transfers` marginally more complicated (no longer a one liner), but I can see what everything is doing based on the names without having to go digging into the implementation.

Aside from the misleading name this looks good though 👍

RndName replied:
i am so in love with these generators but i agree for the sake of readability i changed it as you said :) it's now even more clearer with `is_completable`

DanAlbert commented on game/transfers.py:
```diff
@@ -166,21 +166,46 @@ def find_escape_route(self) -> Optional[ControlPoint]:
+
+        if not self.position.is_friendly(self.player):
+            logging.info(f"Current transfer position {self.position} was captured.")
+            return False
```
Minor bug here: when this is called by `perform_transfers` it doesn't matter if `self.position` is friendly or not. If the base they are _leaving_ is captured but they have a transport, they can still proceed to the next stop. That may be why this was originally handled by `proceed` rather than by `perform_transfers`?

RndName replied:
originally it was never handled thats why i made this PR :) the handling was just checking for the destination rather than checking the path.
This led to a `ValueError` as `shortest_path_with_cost` checked if the `origin` and `destination` are in the **same network**. which they are not as origin was captured -> Thats also the answer to the question above what was the intention of this bug fix.

I will have a look on this one again. At least then it will reach 100% perfectness :)

RndName replied:
i de-complicated the logic and now only check if the processing can be fullfilled for already processing transfers by checking if the next_stop is friendly. For unplanned transfers i only check if the transit network offers us a route between current position and destination as well as checking if the current position they're halting at is friendly

DanAlbert replied:
> originally it was never handled

Is that right? I think you're right about there being a bug for a cheat capture, but I think a regular capture worked?

The base is captured in `Event.commit_captures()`. Units that are _stationed_ (not in a transfer) at that base are retreated or captured. `Game.pass_turn` runs after, which eventually calls `Coaltiion.end_turn`, which calls `PendingTransfers.perform_tansfers`, which calls `TransferOrder.proceed`. The transfer is only disbanded if the _destination_ was captured. If the current location was captured it will still proceed.

The changes to check whether the path is completable a good addition 👍

DanAlbert commented on game/transfers.py:
```diff
@@ -166,21 +166,46 @@ def find_escape_route(self) -> Optional[ControlPoint]:
+                    f"Next stop of transfer {self.next_stop} was captured. "
+                    f"Rerouting the transfer."
+                )
+                self.transport = None
```
A function named `is_completable` shouldn't be unassigning transports. If this needs to move out of `proceed`, it should probably be in `perform_transfers`?

RndName replied:
true! not needed anymore at all ;)

DanAlbert commented on game/transfers.py:
```diff
@@ -166,21 +166,46 @@ def find_escape_route(self) -> Optional[ControlPoint]:
+
+        if self.transport is not None and not self.next_stop.is_friendly(self.player):
+            # Transports
+            if network.has_link(self.position, self.destination):
```
`has_link` returns true if there's a _direct_ link between the two nodes. This can return False when a _path_ exists. You want `has_path_between`.

RndName replied:
must have overseen the small but huge thing that it cheks `is a in nodes[b]` i thought it just checks if a and b are in nodes. but as i am thinking more and lookin up what `has_path_between` does i get the point with the airlifts.
I will add some comments to those functions as, tbh, this name was really misleading. I thought this is too simple to look up :/ But thats a good example for the point you mentioned above!

DanAlbert reviewed (changes requested):
I think I've lost sight of which part of this change is the bug fix. Is it the pruning of uncompletable transfers in `plan_transports`, in `plan_transfers`, or is it the check for the friendliness of the next stop that now takes place in `proceed`?

RndName commented:
> 
> 
> I think I've lost sight of which part of this change is the bug fix. Is it the pruning of uncompletable transfers in `plan_transports`, in `plan_transfers`, or is it the check for the friendliness of the next stop that now takes place in `proceed`?

My initial idea was to improve the logic of disbanding the whole group if the destination was captured so that the convoy can proceed so long until it is at the last stop before the destination and then disband if it is still captured. This gives the coalition the time to recapture the point or to help by sending the reinforcements there where they are needed - to the front. Instead the current code let them at the place where they are. This toally makes sense but the user has to take care of it on his own. AI will handle it.

Maybe i was to over encouraged with it because as i am now debugging more the next_stop thing would have never worked for it as the path calculation always and ever takes in account the destination. I will revert it and let is a an enhancement beside of this bug fix

DanAlbert commented:
> > I think I've lost sight of which part of this change is the bug fix. Is it the pruning of uncompletable transfers in `plan_transports`, in `plan_transfers`, or is it the check for the friendliness of the next stop that now takes place in `proceed`?
> 
> My initial idea was to improve the logic of disbanding the whole group if the destination was captured so that the convoy can proceed so long until it is at the last stop before the destination and then disband if it is still captured. This gives the coalition the time to recapture the point or to help by sending the reinforcements there where they are needed - to the front. Instead the current code let them at the place where they are. This toally makes sense but the user has to take care of it on his own. AI will handle it.

Ahhhhh okay. There's some behavior needed here that I'd never gotten to where the AI needs to be smart enough to redeploy back-line units that would handle that. It would see the unused units and plan a new transfer which would make that unnecessary. What you're suggesting is definitely a better interim solution!

> Maybe i was to over encouraged with it because as i am now debugging more the next_stop thing would have never worked for it as the path calculation always and ever takes in account the destination. I will revert it and let is a an enhancement beside of this bug fix

Yeah, that's trickier. Should this be converted to a draft then? I'm going to have a quick look either way because there's plenty of good refactoring here as well. I was asking mainly because I was trying to figure out where the behavior change was supposed to be :)

RndName commented:
> > > I think I've lost sight of which part of this change is the bug fix. Is it the pruning of uncompletable transfers in `plan_transports`, in `plan_transfers`, or is it the check for the friendliness of the next stop that now takes place in `proceed`?
> > 
> > 
> > My initial idea was to improve the logic of disbanding the whole group if the destination was captured so that the convoy can proceed so long until it is at the last stop before the destination and then disband if it is still captured. This gives the coalition the time to recapture the point or to help by sending the reinforcements there where they are needed - to the front. Instead the current code let them at the place where they are. This toally makes sense but the user has to take care of it on his own. AI will handle it.
> 
> Ahhhhh okay. There's some behavior needed here that I'd never gotten to where the AI needs to be smart enough to redeploy back-line units that would handle that. It would see the unused units and plan a new transfer which would make that unnecessary. What you're suggesting is definitely a better interim solution!
> 
 
I skipped this additional step now and reverted to the behaviour of disbanding the transfer at the current step. The ai will in the next round order a new transfer to the front either way so I think my idea was to complicated at all :D 
> > Maybe i was to over encouraged with it because as i am now debugging more the next_stop thing would have never worked for it as the path calculation always and ever takes in account the destination. I will revert it and let is a an enhancement beside of this bug fix
> 
> Yeah, that's trickier. Should this be converted to a draft then? I'm going to have a quick look either way because there's plenty of good refactoring here as well. I was asking mainly because I was trying to figure out where the behavior change was supposed to be :)

it is not necessary to make it a draft as I reverted all the additional things and only focused on handling the cases which were net already handled and caused the initial failure. So this PR focuses only on the necessary parts :)

DanAlbert reviewed (approved):
Thanks for sticking with it. This looks right 👍
